src/test/java/com/arpnetworking/metrics/common/sources/KafkaSourceTest.java

```diff
@@ -99,102 +124,208 @@ public void testSourceMultiWorkerSuccess() {
+        _source = new KafkaSource<>(new KafkaSource.Builder<String, String>()
+                .setName("KafkaSource")
+                .setConsumer(createMockConsumer(expectedConsumerRecords()))
+                .setParser(new StringParser())
+                .setPollTime(POLL_DURATION)
+                .setNumWorkerThreads(1)
+                .setPeriodicMetrics(_periodicMetrics),
+                new FillingBlockingQueue(bufferSize));
+    }
+
+    private void createExceptionSource(final Exception exception, final boolean onlyOnce) {
+        final Consumer<String, String> consumer = Mockito.spy(createMockConsumer(ConsumerRecords.empty()));
+
+        if (onlyOnce) {
+            Mockito.when(consumer.poll(Mockito.any()))
+                    .thenThrow(exception)
+                    .thenReturn(expectedConsumerRecords())
+                    .thenCallRealMethod();
+        } else {
+            Mockito.when(consumer.poll(Mockito.any()))
+                    .thenThrow(exception);
+        }
+
         _source = new KafkaSource<>(new KafkaSource.Builder<String, String>()
                 .setName("KafkaSource")
                 .setConsumer(consumer)
                 .setParser(new StringParser())
-                .setPollTime(POLL_DURATION),
+                .setPollTime(POLL_DURATION)
+                .setPeriodicMetrics(_periodicMetrics)
+                .setBackoffTime(Duration.ofMillis(10)),
                 _logger);
     }
 
-    private void createBadParsingSource() throws ParsingException {
-        final Parser<String, String> parser = Mockito.mock(StringParser.class);
+    private void createBadParsingSource(final ParsingException exception) throws ParsingException {
+        final Parser<String, String> parser = Mockito.spy(new StringParser());
         Mockito.when(parser.parse(Mockito.anyString()))
-                .thenThrow(new ParsingException("Could not parse data", "bad_data".getBytes(Charsets.UTF_8)));
+                .thenThrow(exception)
+                .thenCallRealMethod();
 
         _source = new KafkaSource<>(new KafkaSource.Builder<String, String>()
                 .setName("KafkaSource")
-                .setConsumer(createMockConsumer())
+                .setConsumer(createMockConsumer(expectedConsumerRecords()))
                 .setParser(parser)
-                .setPollTime(POLL_DURATION),
+                .setPollTime(POLL_DURATION)
+                .setPeriodicMetrics(_periodicMetrics),
                 _logger);
     }
 
+    private void assertMetricCountsCorrect(final long inRecords, final long outRecords, final long parsingExceptions,
+                                           final long kafkaExceptions, final long consumerExceptions) {
+        _periodicMetrics.run(); // Flush the metrics
+        final List<Long> expectedInRecords =
+                _periodicMetrics.getCounters(_source._recordsInCountMetricName) == null
+                ? Collections.singletonList(0L)
```

Comment:
I would remove the burden of these checks by documenting/annotating that `CollectorPeriodicMetrics#getCounters` always returns a non-null list and just return a `Collections.emptyList()` by default. 

This way all of these ternary expressions can go away since the sum of an empty list is still 0.